llvm/lib/ExecutionEngine/JITLink/XCOFFLinkGraphBuilder.cpp

```diff
@@ -0,0 +1,420 @@
+
+    auto Flags = Symbol.getFlags();
+    if (!Flags)
+      return Flags.takeError();
+
+    bool External = *Flags & object::SymbolRef::SF_Undefined;
+    bool Weak = *Flags & object::SymbolRef::SF_Weak;
+    bool Hidden = *Flags & object::SymbolRef::SF_Hidden;
+    bool Global = *Flags & object::SymbolRef::SF_Global;
+
+    auto SymbolIndex = Obj.getSymbolIndex(Symbol.getEntryAddress());
+    auto SymbolName = Symbol.getName();
+    if (!SymbolName)
+      return SymbolName.takeError();
+
+    if (External) {
+      LLVM_DEBUG(dbgs() << "      created external symbol\n");
+      SymbolIndexTable[SymbolIndex] =
+          &G->addExternalSymbol(*SymbolName, Symbol.getSize(), Weak);
+      continue;
+    }
+
+    if (!Symbol.isCsectSymbol()) {
+      LLVM_DEBUG(dbgs() << "      skipped: not a csect symbol\n");
+      continue;
+    }
+
+    auto ParentSym = getXCOFFSymbolContainingSymbolRef(Obj, Symbol);
+    object::XCOFFSymbolRef CsectSymbol = ParentSym ? *ParentSym : Symbol;
+
+    auto CsectSymbolIndex = Obj.getSymbolIndex(CsectSymbol.getEntryAddress());
+    auto ParentSectionNumber = CsectSymbol.getSectionNumber();
+
+    bool IsUndefinedSection = !SectionTable.contains(ParentSectionNumber);
+    Section *ParentSection = !IsUndefinedSection
+                                 ? SectionTable[ParentSectionNumber].Section
+                                 : UndefSection;
+    Block *B = nullptr;
+
+    // TODO: Clean up the logic for handling undefined symbols
+    if (!CsectTable.contains(CsectSymbolIndex) && !IsUndefinedSection) {
+      object::SectionRef &SectionRef =
+          SectionTable[ParentSectionNumber].SectionData;
+      auto Data = SectionRef.getContents();
+      if (!Data)
+        return Data.takeError();
+      auto CsectSymbolAddr = CsectSymbol.getAddress();
+      if (!CsectSymbolAddr)
+        return CsectSymbolAddr.takeError();
+
+      ArrayRef<char> SectionBuffer{Data->data(), Data->size()};
+      auto Offset = *CsectSymbolAddr - SectionRef.getAddress();
+
+      LLVM_DEBUG(dbgs() << "      symbol entry: offset = " << Offset
+                        << ", size = " << CsectSymbol.getSize()
+                        << ", storage class = "
+                        << getStorageClassString(CsectSymbol.getStorageClass())
+                        << "\n");
+
+      B = &G->createContentBlock(
+          *ParentSection, SectionBuffer.slice(Offset, CsectSymbol.getSize()),
+          orc::ExecutorAddr(*CsectSymbolAddr), CsectSymbol.getAlignment(), 0);
+
+      CsectTable[CsectSymbolIndex] = B;
+    } else {
+      B = CsectTable[CsectSymbolIndex];
+    }
+
+    Scope S{Scope::Default};
+    if (Hidden)
+      S = Scope::Hidden;
+    // TODO: Got this from llvm-objdump.cpp:2938 not sure if its correct
+    if (!Weak) {
+      if (Global)
+        S = Scope::Default;
+      else
+        S = Scope::Local;
+    }
+
+    // TODO: figure out all symbols that should have Scope::SideEffectsOnly
```

Comment:
SideEffectsOnly is special, and closely related to the _initializer symbol_ (which should always be `SideEffectsOnly`):

A graph should have an _initializer_ symbol if there are any static constructors in the graph (or related registrations, like ObjC / Swift metadata registrations on Darwin) that should run at _dlopen_ time.

A side-effects only symbol is one that can be looked up to trigger linking of a graph, but which does not show up in the final symbol table (so it must always be looked up with `WeaklyReferencedSymbol` to avoid a `SymbolsNotFound` error). The initializer symbol should be marked with `SideEffectsOnly`, but API clients (e.g. plugins) can add their own special symbols if they want.